### ISSUE TITLE
"Set Selenium Speed" doesn't work when called before opening browser in release 3.0.0b1

### ISSUE THREAD

**Issue by hidebear** [robotframework/SeleniumLibrary#898]
Hello,
I have some test cases wrote by RF script and Selenium2Library(1.8.0) and worked fine.
However after upgrading Selenium2Library to 3.0.0b1, the keyword "Set Selenium Speed" doesn't work anymore, operating the browser by selenium becomes so fast that cause my original test cases failed.

Using keyword "Get Selenium Speed" after setting it, the result is the one I set it:
>${Speed} = Selenium2Library.Get Selenium Speed
>INFO 	${Speed} = 0 seconds
>Selenium2Library.Set Selenium Speed 0.3 s
>${Speed} = Selenium2Library.Get Selenium Speed
>INFO 	${Speed} = 300 milliseconds

But observing the real behaviour on the browser, seems the speed is 0.

Therefore I need to downgrade Selenium2Library back to 1.8.0, and everything back to normal, browser operating speed is actually what I set.

**Environment:**
1. OS: Windows 7 64 bits
2. Python 2.7.11
3. Original Selenium2Library: 1.8.0
4. New Selenium2Library: 3.0.0b1
5. Browser: Chrome 61.0.3163.79 (64 bits)
6. Web driver: ChromeDriver 2.32.498550

**Comment by aaltat** [comment]
This is possible because this part was refactored when library new architecture was created. But there are two [acceptance test](https://github.com/robotframework/SeleniumLibrary/blob/master/test/acceptance/keywords/set_selenium_speed.robot) which try to ensure that `Set Selenium Speed` works correctly.

Would you be able to demonstrate the problem whit a simple test case which we could also run?

**Comment by pekkaklarck** [comment]
This definitely could be broken after all changes to the code. As @aaltat pointed out, there are tests for this, though, and I couldn't reproduce this with a simple example either when using 3.0 beta 1 either.

Tatu, the second test you linked to has, in my opinion, overly long delay slowing down execution. Could easily save like 4 seconds there, but I'm not sure does it really matter in the big picture.

**Comment by aaltat** [comment]
Yes the timeout is too long, usually the Selenium commands happen in fraction of the second. I will change the timeout to something smaller and test it with sauce labs too.

**Comment by hidebear** [comment]
I just did a simple test about typing keyword and click search button on Google search page.

It was tested on both Win7 64bits and Win10 64 bits and the issue is reproducible.

below are the steps:

`(step 3 is the result ran with 1.8.0, executing time = 23 seconds)`
`(step 5 is the result ran with 3.0.0b1, executing time = 1 seconds)`

1) Here is test case: 
[SetSeleniumSpeedTest.zip](https://github.com/robotframework/SeleniumLibrary/files/1309621/SetSeleniumSpeedTest.zip)
2) First check original robotframework-selenium2library:
> C:\Users\user>python -m robot.libdoc Selenium2Library version
1.8.0
3) Ran the test with command "pybot SetSeleniumSpeedTest.txt", the result is:
>${speed} = Selenium2Library.Get Selenium Speed
>INFO   ${speed} = 0 seconds
>Selenium2Library.Set Selenium Speed 3
>${speed} = Selenium2Library.Get Selenium Speed
>INFO   ${speed} = 3 seconds
>Selenium2Library.Open Browser https://www.google.com.tw/?hl=en-US, chrome
>${start} = BuiltIn.Get Time epoch
>...
>Do type keyword, press tab key and click search button
>...
>${end} = BuiltIn.Get Time epoch
>${total} = BuiltIn.Evaluate ${end} - ${start}
>BuiltIn.Log Executing time = ${total} secs.
>INFO   Executing time = 23 secs.
>Selenium2Library.Close All Browsers
4) Upgrade robotframework-selenium2libary to 3.0.0b1
>C:\Users\user>pip install --upgrade --pre robotframework-selenium2library
>Collecting robotframework-selenium2library
>  Downloading robotframework-selenium2library-3.0.0b1.tar.gz (68kB)
>    100% |████████████████████████████████| 71kB 166kB/s
>Collecting selenium>=2.53.6 (from ...
>...
>C:\Users\user>python -m robot.libdoc Selenium2Library version
>3.0.0b1
5) Ran the test again with command "pybot SetSeleniumSpeedTest.txt", the result is:
>${speed} = Selenium2Library.Get Selenium Speed
>INFO   ${speed} = 0 seconds
>Selenium2Library.Set Selenium Speed 3
>${speed} = Selenium2Library.Get Selenium Speed
>INFO   ${speed} = 3 seconds
>Selenium2Library.Open Browser https://www.google.com.tw/?hl=en-US, chrome
>${start} = BuiltIn.Get Time epoch
>...
>Do type keyword, press tab key and click search button
>...
>${end} = BuiltIn.Get Time epoch
>${total} = BuiltIn.Evaluate ${end} - ${start}
>BuiltIn.Log Executing time = ${total} secs.
>INFO   Executing time = 1 secs.
>Selenium2Library.Close All Browsers

**Comment by pekkaklarck** [comment]
I can reproduce the problem with the provided example. I also noticed that the problem occurs if you call `Set Selenium Speed` before opening the browser, but not if you first open the browser and only then set the speed. Can you @hidebear test do you see the same if you change the order?

Things we need to do:

1. Check how the old and new code actually differ in this regard. Did the old implementation affect all browsers and does the new implementation affect only the active browser? Which behavior do we prefer?

2. If we decide that `Set Selenium Speed` should affect only the active browser, then it's logical that using it before `Open Browser` doesn't work. It shouldn't fail to work silently, though, but instead there should be a failure or at least a warning.

**Comment by aaltat** [comment]
I remember making the change, but I didn't see this one coming. In the version 1.8.0 we did always monkey patch the driver during the driver creation. Therefore `Set Selenium Speed` works also before creating the browser/driver. In release 3.0.0  that monkey patch is not done by default, instead it's done only when the `Set Selenium Speed` is called.

This change causes this problem and also one another problem. If I do this: 
```
| Open browser |
| Set Selenium Speed | 5 |
| Open browser |
```
Then the speed is only applied to the first browser and not to the second browser in 3.0.0 release. But in 1.8.0 release it was applied to both browsers.

For me it would feel more logical apply the speed only to the current open browsers, although it's backwards incompatible.

**Comment by aaltat** [comment]
In 1.8.0 this was done always: https://github.com/robotframework/SeleniumLibrary/blob/v1.8.0/src/Selenium2Library/webdrivermonkeypatches.py

But in 3.0.0 it's only done in here: https://github.com/robotframework/SeleniumLibrary/blob/master/src/SeleniumLibrary/keywords/browsermanagement.py#L707

**Comment by pekkaklarck** [comment]
After thinking this a bit more, I think we have three possibilities:

1. Set Selenium Speed affects all browsers. This is same as in earlier versions so no backwards incompatibility (bwic) concerns.

2. Set Selenium Speed affects only the active browser. This is the current behavior. Trying to use the keyword when no browser is open should cause a failure or warning.

3. Set Selenium Speed affects the active browser, but if there is no browser open, it sets the default speed to use. This would be backwards compatible with earlier when using only one browser, and when using multiple and only setting the speed once. Setting the speed between opening browsers would behave differently, but that ought to be pretty rare.

In my opinion the bwic concerns are pretty small in this kind of a major regardless what solution we use. It's would always be possible to set the speed after opening the browser. The main deciding factor I'd use is how other similar settings work. In other words, are current timeouts and waits per browser or do they affect all the browsers?

**Comment by aaltat** [comment]
Because others timeouts are always global, it would be best to make selenium speed also global.

I will write few tests and provide a PR to fix it.

**Comment by aaltat** [comment]
The #910 provides a fix for the problem.

**Comment by aaltat** [comment]
@hidebear the fix is now merged to the master and if you want you can try it out by installing the latest version master. For that see the instructions from the README: https://github.com/robotframework/SeleniumLibrary/blob/master/README.rst#installation

**Comment by hidebear** [comment]
Hi, I just tried 3.0.0b2.dev1, set selenium speed before opening browser worked fine, thanks.

**Comment by pekkaklarck** [comment]
Regressions compared to 1.8 that are already fixed don't belong to 3.0 release notes.
